File: s.pl
use integer;
@A = split //, <>;

sub R {
    for $i ( 0 .. 80 ) {
        next if $A[$i];
        my %t = map {
                 $_ / 9 == $i / 9
              || $_ % 9 == $i % 9
              || $_ / 27 == $i / 27 && $_ % 9 / 3 == $i % 9 / 3
              ? $A[$_]
              : 0 => 1
        } 0 .. 80;
        R( $A[$i] = $_ ) for grep { !$t{$_} } 1 .. 9;
        return $A[$i] = 0;
    }
    die @A;
}
R

File: s.rb
#!/usr/bin/env ruby19

def r
  (0..80).each { |i|
    next if A[i] > 0
    ((1..9).to_a - (0..80).map { |j|
      (j/9 == i/9 ||
        j%9 == i%9 ||
        j/27 == i/27 && j%9/3 == i%9/3) ? A[j] : 0
    }).each { |g|
      A[i] = g
      r
      A[i] = 0
    }
    return
  }
  puts A.join
  exit
end
A = gets.strip.split(//).map {|n| n.to_i }
r
